added section on tracebacks

iritkatriel opened the pull request:
Added a section on traceback groups.

I assume there will be a section on exception group API before it, so the reference to splitting operations etc make sense. 

We'll need some pictures of trees.

iritkatriel commented on except_star.md:
```diff
@@ -483,6 +483,99 @@ errors is error prone.
+for k in keys:
+    if not isinstance(eg.__traceback__, TracebackGroup):
+        # simple traceback, convert it to a group    (WILL WE EVER NEED THIS?)
+        eg.__traceback__ = TracebackGroup.fromTraceback(eg.__traceback__)
```
I'm not at all sure that this 'TracebackGroup.fromTraceback' call can ever happen - I put it in for completion, so that we can hopefully rule it out.

Why would we add an exception to an ExceptionGroup that was created in another frame? Wouldn't that mean that we create a new ExceptionGroup in the current frame and insert our exception along with the original exception group, so that we end up with the original EG nested in the new one?

iritkatriel replied:
Maybe we should first write down the spec for ExceptionGroup and how it is manipulated during a try-except* block. Then we will know what invariants we can assume for the TracebackGroup operations.

1st1 replied:
> Maybe we should first write down the spec for ExceptionGroup and how it is manipulated during a try-except* block. 

+1. It's the ExceptionGroup API and how they're used by try..except* that will guide the TracebackGroup design, not the other way around.

iritkatriel commented on except_star.md:
```diff
@@ -483,6 +483,99 @@ errors is error prone.
+tb = e.__traceback__
+while tb:
+    if isintance(tb, TracebackGroup):
+        next = tb.tb_next_map[e]
```
I think it shouldn't be possible if the data structure is intact. Each exception should lead to a leaf of the tree.

iritkatriel commented on except_star.md:
```diff
@@ -483,6 +483,99 @@ errors is error prone.
+To **split out a subset** of the exceptions in an `ExceptionGroup` that
+match a certain except* clause, we create a new empty `ExceptionGroup`,
+and then perform a sequence of deletions from the original group and
+insertions into the new group. These operations update the tracebacks
+as well:
```
It's worth considering. Does that impact the algorithms much?

gvanrossum replied:
Probably not much. Dealing with the tracebacks will be the most complicated part either way.

1st1 commented on except_star.md:
```diff
@@ -483,6 +483,99 @@ errors is error prone.
+The `TracebackGroup` type is a subclass of `Traceback` and therefore can be
+assigned to the `__traceback__` field of an exception.
+Instead of a single `tb_next` link to a single traceback, the
+`TracebackGroup` has a `tb_next_map` that maps each exception from the
```
Looks like it's a `WeakKeyDict`.

gvanrossum commented on except_star.md:
```diff
@@ -483,6 +483,30 @@ errors is error prone.
+in a traceback's list of frames are from the oldest to the newest frame.
+Appending a new frame is then simply a matter of inserting a new head to the
+linked list referenced from the exception's `__traceback__` field. Crucially,
+the traceback is immutable in the sense that once frames are added they are
+no longer modified.
```
It's really the singly-linked list that's no longer modified, right? So the list is immutable.

gvanrossum reviewed (commented):
I am so glad we don't need a separate tree of tracebacks! I think this is an important simplification of the design. (I should have realized it was too complicated when I tried to draw diagrams of how it worked. :-)

I left two nits to the text.

iritkatriel commented:
I have a concern re trackbacks though. We said an ExceptionGroup is iterable so list(eg) is a flat list of the contained exceptions. But the trackbacks of those exceptions are partial. Their cause and context are wrong too, they should be the same as those of the topmost eg.  It seems that we need an operation to extract an exception with the correct tb, cause, context. The tb part of the eg needs to be copied. Then something around the iteration api, not sure what. Or just document that the metadata is not reconstructed when you take an exception out of a group.

iritkatriel commented:
We sort of do have that extract operation already - when we extract an exception in the form of a singleton ExceptionGroup via project() then the group has the correct metadata.

1st1 commented on except_star.md:
```diff
@@ -642,6 +666,16 @@ to start using the new `except *` syntax right away.  They will have to use
+### Traceback Representation
+
+We considered options for adapting the traceback data structure to represent
+trees, but it became aparent that a traceback tree is not meaninful once separated
```
```suggestion
trees, but it became apparent that a traceback tree is not meaningful once separated
```

1st1 reviewed (approved):
LGTM. Happy we can keep tracebacks simple!

1st1 commented:
> Then something around the iteration api, not sure what. Or just document that the metadata is not reconstructed when you take an exception out of a group.

IMO just documenting that would be OK. We can also add some helpers to the `traceback` module to flatten ExceptionGroup objects more carefully.

iritkatriel commented:
> > Then something around the iteration api, not sure what. Or just document that the metadata is not reconstructed when you take an exception out of a group.
> 
> IMO just documenting that would be OK. We can also add some helpers to the `traceback` module to flatten ExceptionGroup objects more carefully.

Maybe we don't need ExceptionGroup to be iterable at all. The use case of filtering OSError by of selecting OSError by code (https://github.com/python/exceptiongroups/blob/master/except_star.md#raising-exceptiongroups-manually) can be implemented with something like `eg.project(lambda e: isinstnace(e, OSError) and e.errno != errno.EPIPE)`.

Is there another use case? 

How about we provide this ``project`` method along with something that squashes singleton EGs, so if we have ``EG("1", SyntaxError(1), EG("2", ValueError(2)))`` then the squashed version is ``EG("1", SyntaxError(1), ValueError(2))`` where the traceback of the EG("2") that was squashed away was concatenated as the prefix of the ValueError is was wrapping.   These two operations are sound and will probably be enough to do anything someone needs to do with ExceptionGroups.

Note that the experimental implementation doesn't have the iterate operation, only a flatten helper method in the unit tests (it's not needed other than for the tests).

1st1 commented:
Can you add a link to the experimental implementation to the README?

gvanrossum commented:
> Maybe we don't need ExceptionGroup to be iterable at all.

Agreed, I have been wondering this for a while. Then we wouldn't have to worry about whether an EG with no exceptions is truthy or falsy or a forbidden edge case.

You can still access the bare subexceptions using `eg.excs` (hm, maybe a less abbrev name here?).

Related, I also wonder whether the constructor should take a list of exceptions instead of being varargs. I.e., `ExceptionGroup("msg", [e1, e2, e3])` instead of `ExceptionGroup("msg", e1, e2, e3)`. Usually the list of subexceptions is already available as a list anyway.

1st1 commented:
> Related, I also wonder whether the constructor should take a list of exceptions instead of being varargs. I.e., 

+1 for the list option. I vaguely recall we already discussed this somewhere else, btw.

iritkatriel commented:
I've added the link in readme. 

Re list - yes, it was a list initially, then we moved to varargs (before we added the msg arg).  Do we want both varargs and sequence? Or just sequence? Or just list?

1st1 commented:
> I've added the link in readme.

Thanks! I'll play with it on the weekend.

> Re list - yes, it was a list initially, then we moved to varargs (before we added the msg arg). Do we want both varargs and sequence? Or just sequence? Or just list?

My preference would be to go with an iterable/sequence (but no varargs).

iritkatriel commented:
I'm thinking about how to implement raise inside an except* block and came across this question:

At the moment, if we do this:

```
try:
    raise ValueError(12)
except *TypeError as e:
    raise
```

we get 

```
Traceback (most recent call last):
  File "C:\Users\User\src\cpython\tmp.py", line 7, in <module>
    raise ValueError(12)
ValueError: 12
```

which is what you would expect when an exception is not matched.


But if we do this:

```
try:
    raise ValueError(12)
except *ValueError as e:
    raise
```
we get:

```
Running Release|x64 interpreter...
Traceback (most recent call last):
  File "C:\Users\User\src\cpython\tmp.py", line 7, in <module>
    raise ValueError(12)
ExceptionGroup
This exception has 1 sub-exceptions:
 ----------------- 1/1 -----------------
 Traceback (most recent call last):
   File "C:\Users\User\src\cpython\tmp.py", line 7, in <module>
     raise ValueError(12)
 ValueError: 12
 ----------------- end of 1 -----------------
```

This is because the except* wraps the ValueError in an exception group, and that gets matched and reraised. 
Is this what we want?

gvanrossum commented:
I think it's acceptable. If you were to simplify the exception if the EG has exactly one subexception, you'd have the converse inconsistency, and I don't think adding more hidden state to track this case is a great idea either.

Bare `raise` by itself already appears to use hidden state to distinguish itself from `raise e`, and that hidden state is already more complex that you'd think because the bare `raise` could occur inside a function called at that point too. I don't want to complicate things further.
